### ISSUE TITLE
C API / shared library for the runtime

### ISSUE THREAD

**Issue by jemc** [microsoft/verona#30]
Exciting to see this released! Thanks for all the work so far.

In the interest of integrating with the Verona runtime from another compiler, I aim to wrap the API for C, so that it can be easily invoked from LLVM IR.

[I have already started work on such a wrapper on my fork](https://github.com/jemc/verona/compare/c-api). It's nowhere near finished, but it's a start as I begin to digest the design of the runtime's API and figure out how to properly wrap it.

Is there interest in eventually merging such a C API into this repository after it's reasonably functional and polished up? If so, we could keep this ticket open and I could track my work here before eventually filing the PR.

If not, then we can close this ticket and I suppose I can make my own repo to house that work with this one as a submodule. Obviously I'd prefer to get it integrated here if possible, but I'll understand if you say it's beyond the focus of this repository at this time.

Thanks again!

**Comment by mjp41** [comment]
@jemc We would certainly take this.  I have been meaning to factor out a shared library to call from the interpreter.  The API surface has somewhat evolved as we have added features, and is not as clean as I would like.  I think you doing this would be a good pressure for us to tidy that up.

The two caveats:
* Anything not tested in CI, will get broken.
* Anything tested in CI will probably get broken, as we will almost certainly need to change the API, so nothing you do will be stable.  But we will at least have to fix the tests.

Please do raise questions about the API surface here.  Having a cleaner C++ API might be a necessary first step.

**Comment by jemc** [comment]
Great. I can incorporate some tests as part of the initial PR.

> Please do raise questions about the API surface here. Having a cleaner C++ API might be a necessary first step.

One place where I hit a bit of a snag last night is `Cown::schedule` which takes that Behaviour class as a template argument and constructs it before scheduling it. Since a C caller can't be expected to make a constructable C++ class, I'm not sure how to wrap this. It would be easier if there were an overload of `Cown::schedule` that takes an already-constructed `Action` instance. Do you think this is a feasible change to the C++ API?

**Comment by jemc** [comment]
Here's an example of what I mean for refactoring `Cown::schedule` to accept an already-constructed `Action` instance, from my fork:
```diff
diff --git a/src/rt/sched/cown.h b/src/rt/sched/cown.h
index f94fc6d..8ec9e4f 100644
--- a/src/rt/sched/cown.h
+++ b/src/rt/sched/cown.h
@@ -674,13 +674,6 @@ namespace verona::rt
         1, &cown, std::forward<Args>(args)...);
     }
 
-    /**
-     * Sends a multimessage to the first cown we want to acquire.
-     *
-     * Pass `transfer = YesTransfer` as a template argument if the
-     * caller is transfering ownership of a reference count on each cown to this
-     * method.
-     **/
     template<
       class Behaviour,
       TransferOwnership transfer = NoTransfer,
@@ -693,6 +686,21 @@ namespace verona::rt
       Alloc* alloc = ThreadAlloc::get();
       Behaviour* b = (Behaviour*)alloc->alloc<sizeof(Behaviour)>();
       Action* action = new (b) Behaviour(std::forward<Args>(args)...);
+
+      schedule<transfer>(count, cowns, action);
+    }
+
+    /**
+     * Sends a multimessage to the first cown we want to acquire.
+     *
+     * Pass `transfer = YesTransfer` as a template argument if the
+     * caller is transfering ownership of a reference count on each cown to this
+     * method.
+     **/
+    template<TransferOwnership transfer = NoTransfer>
+    static void schedule(size_t count, Cown** cowns, Action* action)
+    {
+      Alloc* alloc = ThreadAlloc::get();
       Cown** sort = (Cown**)alloc->alloc(count * sizeof(Cown*));
       memcpy(sort, cowns, count * sizeof(Cown*));
```

And here's how I'm wrapping it:

```c++
///
// RTAction

RTAction* RTAction_new(RTActionDescriptor* desc) {
  return reinterpret_cast<RTAction*>(
    new rt::Action(reinterpret_cast<rt::Action::Descriptor*>(desc))
  );
}

void RTAction_schedule(RTAction* action, RTCown** cowns, size_t count) {
  rt::Cown::schedule(
    count,
    reinterpret_cast<rt::Cown**>(cowns),
    reinterpret_cast<rt::Action*>(action)
  );
}
```

Note that in the wrapper I pretended it was a method of `Action` instead of `Cown`, because that made more ergonomic sense to me (though from a code organization perspective, I can understand why you have it in `cown.h` instead of `action.h`.

**Comment by jemc** [comment]
Okay, I've got a basic equivalent of the `bank1.verona` example working, except the memory management is still broken. I must admit I don't yet understand under what circumstances I am supposed to call `Cown::acquire` and `Cown::release`.

Here's my example code:
https://github.com/jemc/verona/blob/2ab6c3711bac01d6f460945c6017ec5aa3c255ee/src/rt_c/verona_c_test.c#L73-L86
```c
  RTScheduler_init(1);
  RTScheduler_set_detect_leaks(true);
  printf("RTScheduler setup complete!\n");

  BankAccount* account = BankAccount_create(1000);
  // RTCown_acquire((RTCown*)account);
  BankAccount_add(account, 200);
  BankAccount_add(account, 500);
  BankAccount_add(account, 1500);
  RTCown_release((RTCown*)account);

  printf("RTScheduler running...\n");
  RTScheduler_run();
  printf("RTScheduler finished!\n");
```

With the explicit `RTCown_release` in that snippet in place, I seem to be getting a double-free, if I interpret the lldb trace correctly.

If I comment-out the `RTCown_release` line, it seems to be a never-free / memory leak. If I do that and also invoke `RTScheduler_set_detect_leaks(false)`, then everything "works".

If possible, I'd appreciate some guidance on how/when acquire/release should be used, both in my example and in general.

**Comment by mjp41** [comment]
@jemc So the schedule code has been written in a specific way to allow the number of allocations for a multimessage to be optimised.  I want to refactor the multimessage, to placement new the payload and the list of cowns into a special runtime object. (Not a standard C++ object).  So I wouldn't take the refactoring you propose as it restricts the runtime layout.

The runtime API has been designed with particular codegen in mind, rather than to expose to a programmer directly. On Monday, I will also go through your API to help specify where Acquires and Releases are needed. 

I will also write some documentation about debugging and tracing on Monday, so that you can see what the runtime is doing.

**Comment by jemc** [comment]
> The runtime API has been designed with particular codegen in mind, rather than to expose to a programmer directly

Right, the same is true of my case - I'm not intending for a end-programmer to use these API calls. I'm just wrapping it in a C API so that my codegen pass can write LLVM IR that interacts with it, without worrying about C++ munging.

**Comment by orcmid** [comment]
I am missing something.  Not all that important within the project, but curious to know motivation.

1. The use of *.h very heavily (yay preprocessor munging) in interesting ways.  I sort of get that.  But these are **_very_** C++ specific and I would have expected to see *.hpp.  It also suggests the motivation for using C++ at this level.
2. Same question about *.cc.  Looks like *.cpp to me, in the few places I examined it (i.e., tests).  

This seems to be very disciplined, and that appeals to me.  I just have this nomenclature niggle.  Is it explained somewhere?

**Comment by mjp41** [comment]
@jemc here is some feedback:

So the following shouldn't be exposed in verona_c.h :

```C++
void RTCown_reschedule(RTCown* cown);

void RTScheduler_record_inflight_message();
void RTScheduler_recv_inflight_message();
bool RTScheduler_no_inflight_messages();
```
They are all internal to the runtime.

The following you might want to expose:
```C++
 RTScheduler_run_with_startup
 RTScheduler_want_ld
```

There is a leak detection protocol this can be used for debugging cycles in the cown graph. `want_ld` is used to trigger that starting.

`run_with_startup` guarantees when the initial threads are added to the runtime, they all run the code supplied.  This means you can set up per thread code and not need initialisation checks.  The VM in the interpreter does this.

**Comment by mjp41** [comment]
Can you confirm that lines like 
```C++
std::atomic<size_t> weak_count = 1;
```
are being run by your API.  I don't think this will be called, and this means a bunch of things are not getting set up correctly.

The `alloc` on `Cown` is doing enough to convert a C++ object allocation into a Verona object one.  But we are still relying on the C++ constructors running.

**Comment by mjp41** [comment]
@orcmid Not sure there is a reason, we used `.cc` for C++ files, and we use `.h` for headers.  We should perhaps of used `.hh`, but the project is pretty much pure C++.

**Comment by orcmid** [comment]
Understood, thanks.  I was just surprised that existing conventions were not used.  It is often the case that .h headers that have C Language (not C++) signatures and library dependencies are crafted to work in compilations of either language, and the .hpp headers (assuming someone is paying attention) are ones that require C++, such as yours.  So a C Language wrapper would have a .h header :).  

The new FAQ explaining that this is a C++ project at the current level makes it clear though, and I suppose VS 2019 is indifferent (or you have options as required -- I am being lazy and not looking).

**Comment by jemc** [comment]
Thanks for the feedback!

> the following shouldn't be exposed in `verona_c.h` [...] They are all internal to the runtime.

Yeah, no worries. I figured that not all of these functions I wrapped would/should be part of the final API contract. I basically wrote a wrapper (or a TODO comment) for any function that was `public` (as opposed to `protected`/`private`) on those classes that I looked at. It was a way for me to familiarize myself with the codebase. I expect that before I do my pull request everything unnecessary can be trimmed away leaving only those parts that are usable and testable.

I have removed the functions you mentioned from my branch, and am of course open to trimming out any other functions in the future that shouldn't be part of the contract.

> The following you might want to expose: `RTScheduler_run_with_startup`, `RTScheduler_want_ld`

Thanks for the tip. I have now wrapped these and added them to be used my example program. However, it still segfaults currently, even after making sure to do all my Cown- and Action-related operations from within the `run_with_startup` function.

> Can you confirm that lines like `std::atomic<size_t> weak_count = 1;` are being run by your API. I don't think this will be called, and this means a bunch of things are not getting set up correctly.

You are correct - this doesn't seem to be running. I used `lldb` to check that the `weak_count` for a Cown returned by `Cown::alloc` is zero rather than one.

I'm not well-versed on C++ arcana, so I'm not sure how to properly make sure this field initialization code executes from within my wrapping code, which as you mention, currently just calls `Cown::alloc` - I'm open to whatever approach you can suggest for that.

For now, I have added a statement of `this->weak_count = 1;` to the top of `Cown::init` as a temporary workaround for this specific case, though I'm expecting you'll likely tell me this isn't the desired approach. Again, I'm open to whatever guidance you have one that.

I also looked for other cases within `Cown` and `Action` that might have the same problem, but did not find any other fields initialized in this way. There are many such fields in `Scheduler`, but I'm assuming this isn't a problem for me, since I'm only using the static global instance of it and not constructing one myself.

---

Note that even after adding the above workarounds to fix initialization of this one field I am still seeing a segfault on my example program. However, it has a different backtrace than it did before, so we seem to be getting somewhere. Thanks again for that.

Here's the output I see in `lldb` now:
```
* thread #2, name = 'verona_rt_c_tes', stop reason = signal SIGSEGV: invalid address (fault address: 0x0)
    frame #0: 0x000000000043a691 verona_rt_c_test`std::__atomic_base<snmalloc::Remote*>::operator snmalloc::Remote*() const [inlined] std::__atomic_base<snmalloc::Remote*>::load(__m=memory_order_seq_cst, this=0x0000000000000000) const at atomic_base.h:740:24
   737 		__glibcxx_assert(__b != memory_order_release);
   738 		__glibcxx_assert(__b != memory_order_acq_rel);
   739 	
-> 740 		return __atomic_load_n(&_M_p, int(__m));
   741 	      }
   742 	
   743 	      _GLIBCXX_ALWAYS_INLINE __pointer_type
(lldb) bt
* thread #2, name = 'verona_rt_c_tes', stop reason = signal SIGSEGV: invalid address (fault address: 0x0)
  * frame #0: 0x000000000043a691 verona_rt_c_test`std::__atomic_base<snmalloc::Remote*>::operator snmalloc::Remote*() const [inlined] std::__atomic_base<snmalloc::Remote*>::load(__m=memory_order_seq_cst, this=0x0000000000000000) const at atomic_base.h:740:24
    frame #1: 0x000000000043a67b verona_rt_c_test`std::__atomic_base<snmalloc::Remote*>::operator snmalloc::Remote*(this=0x0000000000000000) const at atomic_base.h:617
    frame #2: 0x0000000000437bc8 verona_rt_c_test`std::atomic<snmalloc::Remote*>::operator snmalloc::Remote*(this=0x0000000000000000) const at atomic:374:16
    frame #3: 0x0000000000434146 verona_rt_c_test`snmalloc::MPSCQ<snmalloc::Remote>::invariant(this=0x0000000000000000) at mpscq.h:24:7
    frame #4: 0x000000000043ab54 verona_rt_c_test`snmalloc::MPSCQ<snmalloc::Remote>::enqueue(this=0x0000000000000000, first=0x00007fffcc000b90, last=0x00007fffcc000b90) at mpscq.h:56:7
    frame #5: 0x0000000000438499 verona_rt_c_test`snmalloc::Allocator<snmalloc::MemoryProviderStateMixin<snmalloc::PALLinux>, snmalloc::DefaultChunkMap<snmalloc::GlobalPagemapTemplate<snmalloc::FlatPagemap<24ul, unsigned char> > >, true, &(snmalloc::lazy_replacement(void*))>::RemoteCache::post(this=0x00007ffff6002668, id=140737320594560) at alloc.h:554:15
    frame #6: 0x0000000000434b53 verona_rt_c_test`snmalloc::Allocator<snmalloc::MemoryProviderStateMixin<snmalloc::PALLinux>, snmalloc::DefaultChunkMap<snmalloc::GlobalPagemapTemplate<snmalloc::FlatPagemap<24ul, unsigned char> > >, true, &(snmalloc::lazy_replacement(void*))>::remote_dealloc(this=0x00007ffff6002000, target=0x0000000000000000, p=0x00007fffcc000b90, sizeclass=1) at alloc.h:1322:7
    frame #7: 0x00000000004302bc verona_rt_c_test`snmalloc::Allocator<snmalloc::MemoryProviderStateMixin<snmalloc::PALLinux>, snmalloc::DefaultChunkMap<snmalloc::GlobalPagemapTemplate<snmalloc::FlatPagemap<24ul, unsigned char> > >, true, &(snmalloc::lazy_replacement(void*))>::dealloc(this=0x00007ffff6002000, p=0x00007fffcc000b90, size=24) at alloc.h:258:11
    frame #8: 0x000000000042ed33 verona_rt_c_test`verona::rt::Cown::run_step(m=0x00007fffd40010c0) at cown.h:661:21
    frame #9: 0x000000000042f043 verona_rt_c_test`verona::rt::Cown::run(this=0x00007fffd4010060, alloc=0x00007ffff6002000, (null)=NotInLD, (null)=EPOCH_A))>*, verona::rt::ThreadState::State, verona::rt::EpochMark) at cown.h:836:22
    frame #10: 0x0000000000439db5 verona_rt_c_test`void verona::rt::SchedulerThread<verona::rt::Cown>::run<void*>(this=0x000000000147feb0, startup=(verona_rt_c_test`startup at verona_rt_c_test.c:68:25), (null)=<unavailable>, args#0=0x0000000000000000)(void*), void*) at schedulerthread.h:261:14
    frame #11: 0x0000000000443c5f verona_rt_c_test`void std::__invoke_impl<void, void (verona::rt::SchedulerThread<verona::rt::Cown>::*)(void (*)(void*), void*), verona::rt::SchedulerThread<verona::rt::Cown>*, void (*)(void*), void*>((null)=__invoke_memfun_deref @ 0x00007fffe4a54dd0, __f=0x0000000001480680, __t=0x0000000001480678, (null)=<unavailable>, (null)=<unavailable>, __args#0=0x0000000001480670, __args#1=0x0000000001480668)(void (*)(void*), void*), verona::rt::SchedulerThread<verona::rt::Cown>*&&, void (*&&)(void*), void*&&) at invoke.h:73:46
    frame #12: 0x0000000000443a27 verona_rt_c_test`std::__invoke_result<void (verona::rt::SchedulerThread<verona::rt::Cown>::*)(void (*)(void*), void*), verona::rt::SchedulerThread<verona::rt::Cown>*, void (*)(void*), void*>::type std::__invoke<void (__fn=0x0000000001480680, (null)=<unavailable>, (null)=<unavailable>, (null)=<unavailable>, __args#0=0x0000000001480678, __args#1=0x0000000001480670, __args#2=0x0000000001480668)(void (*)(void*), void*), verona::rt::SchedulerThread<verona::rt::Cown>*, void (*)(void*), void*>(void (verona::rt::SchedulerThread<verona::rt::Cown>::*&&)(void (*)(void*), void*), verona::rt::SchedulerThread<verona::rt::Cown>*&&, void (*&&)(void*), void*&&) at invoke.h:95:40
    frame #13: 0x0000000000443851 verona_rt_c_test`void std::thread::_Invoker<std::tuple<void (verona::rt::SchedulerThread<verona::rt::Cown>::*)(void (*)(void*), void*), verona::rt::SchedulerThread<verona::rt::Cown>*, void (*)(void*), void*> >::_M_invoke<0ul, 1ul, 2ul, 3ul>(this=0x0000000001480668, (null)=_Index_tuple<0, 1, 2, 3> @ 0x00007fffe4a54e60) at thread:244:26
    frame #14: 0x000000000044376a verona_rt_c_test`std::thread::_Invoker<std::tuple<void (verona::rt::SchedulerThread<verona::rt::Cown>::*)(void (*)(void*), void*), verona::rt::SchedulerThread<verona::rt::Cown>*, void (*)(void*), void*> >::operator(this=0x0000000001480668)() at thread:251:31
    frame #15: 0x0000000000443734 verona_rt_c_test`std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (verona::rt::SchedulerThread<verona::rt::Cown>::*)(void (*)(void*), void*), verona::rt::SchedulerThread<verona::rt::Cown>*, void (*)(void*), void*> > >::_M_run(this=0x0000000001480660) at thread:195:13
    frame #16: 0x00007ffff7e5b6f4 libstdc++.so.6`___lldb_unnamed_symbol615$$libstdc++.so.6 + 20
    frame #17: 0x00007ffff7f914e2 libpthread.so.0`start_thread + 226
    frame #18: 0x00007ffff7b5c693 libc.so.6`__GI___clone + 67
(lldb) 
```

**Comment by mjp41** [comment]
>For now, I have added a statement of this->weak_count = 1; to the top of Cown::init as a temporary workaround for this specific case, though I'm expecting you'll likely tell me this isn't the desired approach. Again, I'm open to whatever guidance you have one that.

That is not a good solution.   There are other classes nested in there, I am pretty certain something has not been initialised and is causing your crash.  I think you need to placement new a cown over the allocation.  
```C++
new (cown) Cown;
```
Now, I think you can put this directly after the `alloc` call into the runtime. The interpreter and unit tests all use `VCown` to build a cown.  That replaces `new` with a call to alloc, but then I believe the constructor is run on what alloc returns.

Sorry for not having more time today, I hope this helps unblock you.

**Comment by jemc** [comment]
Ah, I see. I didn't know I could run a C++ constructor on an already-allocated object. I've updated my branch to do this, reverted the aforementioned `init` function workaround, and confirmed that it now initializes the `weak_count` properly in my `RTCown_alloc` function.

I'm still seeing the same segmentation fault I mentioned above, but I'll spend some time with systematic testing and/or crash logging later today to try to dig deeper into the issue.

Thanks for the help.

**Comment by jemc** [comment]
Okay, I resolved the remaining issue by realizing that the opposite of the above Cown problem was happening with Action.

While with Cown, I was allocating correctly (using `snmalloc` via `Cown::malloc`) but not calling the constructor, with Action I was calling the constructor normally (i.e. using the default allocator instead of `snmalloc`). When I modified `RTAction_new` to be implemented as follows, the test program ran all the way through without issue:
```c++
RTAction* RTAction_new(RTActionDescriptor* desc) {
  rt::Action* action = (rt::Action*)rt::ThreadAlloc::get()->alloc(desc->size);
  new (action) rt::Action(reinterpret_cast<rt::Action::Descriptor*>(desc));
  return reinterpret_cast<RTAction*>(action);
}
```

Now I'm going to focus on trying to make some reasonable test cases for all this so they can be included with CI, then submit the PR for review and further feedback.

In the meantime, though, I still don't feel like I understand when/if a caller of the API should be using `Cown`'s `acquire` and `release` functions (or for that matter, `weak_acquire`, `weak_release`, and `acquire_strong_from_weak` as well). It would be great if we could get some kind of documentation that describes in brief what situations each of these is to be used in. Then I can be sure my compiler is generating the appropriate LLVM IR so that this happens on behalf of the end-user programmer without them having to be aware of it.
